File: cpp/problem017.cpp
/*
 * problem017.cpp
 * 
 * Problem 17: Number letter counts
 *
 * If the numbers 1 to 5 are written out in words: one, two, three, four, five,
 * then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.
 *
 * If all the numbers from 1 to 1000 (one thousand), inclusive were written out
 * in words, how many letters would be used?
 *
 * NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and
 * forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20
 * letters. The use of "and" when writing out numbers is in compliance with
 * British usage.
 * 
 * Author: Curtis Belmonte
 * Created: Aug 29, 2014
 */

#include <iostream>

//#include "common.h"

using namespace std;

/* PARAMETERS ****************************************************************/

// TODO: parameterize upper limit

/* SOLUTION ******************************************************************/

int main() {
    // compute the sum of all digit letter counts, excluding 0
    const unsigned int kAllDigitLetters =
            3 // strlen("one")
          + 3 // strlen("two")
          + 5 // strlen("three")
          + 4 // strlen("four")
          + 4 // strlen("five")
          + 3 // strlen("six")
          + 5 // strlen("seven")
          + 5 // strlen("eight")
          + 4 // strlen("nine")
    ;

    // compute the sum of all "teen" letter counts, including 10, 11, and 12
    const unsigned int kAllTeenLetters =
            3 // strlen("ten")
          + 6 // strlen("eleven")
          + 6 // strlen("twelve")
          + 8 // strlen("thirteen")
          + 8 // strlen("fourteen")
          + 7 // strlen("fifteen")
          + 7 // strlen("sixteen")
          + 9 // strlen("seventeen")
          + 8 // strlen("eighteen")
          + 8 // strlen("nineteen")
     ;

    // compute the sum of all multiples of ten letter counts, excluding 10
    const unsigned int kAllTenLetters =
            6 // strlen("twenty")
          + 6 // strlen("thirty")
          + 5 // strlen("forty")
          + 5 // strlen("fifty")
          + 5 // strlen("sixty")
          + 7 // strlen("seventy")
          + 6 // strlen("eighty")
          + 6 // strlen("ninety")
     ;

    // count the letters of other useful number words
    const unsigned int kAndLetters = 3; // strlen("and");
    const unsigned int kHundredLetters = 7; // strlen("hundred");
    const unsigned int kOneLetters = 3; // strlen("one");
    const unsigned int kThousandLetters = 8; // strlen("thousand");

    // count the letters of all numbers below 20
    unsigned int letter_count = kAllDigitLetters + kAllTeenLetters;

    // count the letters of all numbers from 20 to 99
    const unsigned int kDigitCount = 9;
    const unsigned int kTenCount = 8;
    letter_count += (kAllTenLetters * (kDigitCount + 1))
                    + (kAllDigitLetters * kTenCount);
    const unsigned int kLettersBelow100 = letter_count;

    // count the letters of all numbers from 100 to 999
    const unsigned int kNumbersBelow100 = 99;
    const unsigned int kAndWordCount = kDigitCount * kNumbersBelow100;
    const unsigned int kHundredWordCount = kAndWordCount + kDigitCount;
    letter_count += (kAllDigitLetters * (kNumbersBelow100 + 1))
                    + (kHundredLetters * kHundredWordCount)
                    + (kAndLetters * kAndWordCount)
                    + (kLettersBelow100 * kDigitCount);

    // count the letters of 1000
    letter_count += kOneLetters + kThousandLetters;

    cout << letter_count << endl;
    return 0;
}

File: py/problem_017.py
"""problem_017.py

Problem 17: Number letter counts

If the numbers 1 to 5 are written out in words: one, two, three, four, five,
then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to MAX_NUMBER inclusive were written out in words,
how many letters would be used?

NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and
forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20
letters. The use of 'and' when writing out numbers is in compliance with
British usage.

Author: Curtis Belmonte
"""

import common as com

# PARAMETERS ##################################################################

MAX_NUMBER = 1000 # default: 1000

# SOLUTION ####################################################################

and_letters = len('and')

# letter counts for unique numbers (see common.NUMBER_WORDS)
num_letters = {0: 0}
for number, word in com.NUMBER_WORDS.items():
    num_letters[number] = len(word)

# letter counts for relevant powers of 10
pow10_letters = {
    100: len('hundred'),
    1000: len('thousand'),
    10**6: len('million'),
}


@com.memoized
def count_letters(n):
    """Returns the number of letters in the written representation of the
    natural number n (in compliance with British usage)."""

    count = 0

    # count letters for powers of 10
    pow10_flag = False
    for pow10 in sorted(pow10_letters.keys(), reverse=True):
        if n >= pow10:
            pow10_flag = True

            div, mod = divmod(n, pow10)
            count += count_letters(div) + pow10_letters[pow10]
            n = mod

    # count letters for "and" if necessary
    if pow10_flag and n > 0:
        count += and_letters

    # count letters for tens place (at and above 20)
    if n >= 20:
        ones_digit = n % 10
        count += num_letters[n - ones_digit]
        n = ones_digit
    
    # count letters for ones place
    count += num_letters[n]

    return count


def solve():
    total = 0
    for n in range(1, MAX_NUMBER + 1):
        total += count_letters(n)
    return total


if __name__ == '__main__':
    print(solve())
